src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSession.java

```diff
@@ -210,14 +223,13 @@ public void operationComplete(ChannelFuture channelFuture) throws Exception {
   }
 
   private void startNegotiation(Channel activeChannel) {
-    logger.info("{}: mark session state negotiation");
-    if (enableAuth) {
-      negotiation = new Negotiation(this);
-      negotiation.start();
-    } else {
-      logger.info("{}: mark session state connected");
-      markSessionConnected(activeChannel);
-    }
+    VolatileFields newCache = new VolatileFields();
+    newCache.state = ConnState.CONNECTING;
+    newCache.nettyChannel = activeChannel;
+    fields = newCache;
```

Comment:
I don’t understand why update `fields`？

Reply:
if we don't update `fields`, the nettyChannel is null. We need nettyChannel to send negotiation message.

Reply:
I think you should not modify any ReplicaSession code to implement negotiation. Implement upon it.
Try to think of any other design alternative. Adding features by intrusive modifications to an existing class is the last thing we should do.

ReplicaSession contains pure RPC logic. The negotiation is essentially based on RPC too. So why break this design? Why add responsibility to ReplicaSession?

One approach is to inherit ReplicaSession to SecureSession and implement negotiation in the subclass.